### PR TITLE
DAC6-2746: Allow large file uploads via SDES

### PR DESCRIPTION
**Pull request by koladeadewuyi**

### CODE REVIEW

File: app/services/submission/SDESService.scala [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+    sdesConnector.sendFileNotification(fileNotificationRequest).map {
+      case Right(_) => Right(conversationId)
+      case Left(errorStatus) =>
+        logger.warn(s"Failed to send file with conversation Id [${conversationId.value}] to SDES. Got error status: $errorStatus")
+        Left(SdesSubmissionError(errorStatus))
+    }
+  }
+
+  def createFileNotificationRequest(
```
Comment (oarshad) [inline app/services/submission/SDESService.scala:64]: can be private

File: it/test/wiremock/WireMockHelper.scala [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+object WireMockHelper {
+
+  val wireMockPort: Int = Using(new ServerSocket(0))(_.getLocalPort)
+    .getOrElse(throw new Exception("Failed to find random free port"))
+
+  implicit class MappingBuilderExt(builder: client.MappingBuilder) {
+
+    def withRequestHeaders(headers: Set[(String, String)]): MappingBuilder =
+      headers.foldLeft(builder) { (builder, header) =>
+        val (key, value) = header
```
Comment (oarshad) [inline it/test/wiremock/WireMockHelper.scala:72]: can be inlined above
```
headers.foldLeft(builder) { case (builder, (key, value)) =>
```

File: test/controllers/submission/SubmissionControllerSpec.scala [hunk 3]
```diff
@@ -18,189 +18,134 @@ package controllers.submission
...
-        .thenReturn(Future.successful(true))
-      when(mockReadSubscriptionService.getContactInformation(any[String]())(any[HeaderCarrier](), any[ExecutionContext]()))
-        .thenReturn(Future.successful(Right(responseDetail)))
-      when(mockSubmissionConnector.submitDisclosure(any[NodeSeq](), any[ConversationId])(any[HeaderCarrier]()))
-        .thenReturn(Future.successful(HttpResponse(NO_CONTENT, "")))
-      when(mockXMLValidationService.validate(any[NodeSeq], any[String]))
-        .thenReturn(Right(basicXml))
-      val submission = basicXml
+    "must treat file having size <= maxNormalFileSizeBytes as normal file" in {
+      forAll(arbitrary[SubmissionDetails]) { submissionDetails =>
```
Comment (oarshad) [inline test/controllers/submission/SubmissionControllerSpec.scala:48]:
```
forAll { submissionDetails: SubmissionDetails =>
```
it should pickup the implicit

File: test/controllers/submission/SubmissionControllerSpec.scala [hunk 4]
```diff
@@ -18,189 +18,134 @@ package controllers.submission
...
-
-      val argumentCaptor: ArgumentCaptor[NodeSeq]                      = ArgumentCaptor.forClass(classOf[NodeSeq])
-      val argumentCaptorSubmissionDetails: ArgumentCaptor[FileDetails] = ArgumentCaptor.forClass(classOf[FileDetails])
-      val argumentCaptorConversationId: ArgumentCaptor[ConversationId] = ArgumentCaptor.forClass(classOf[ConversationId])
-
-      verify(mockFileDetailsRepository, times(1)).insert(argumentCaptorSubmissionDetails.capture())
-      verify(mockSubmissionConnector, times(1)).submitDisclosure(argumentCaptor.capture(), argumentCaptorConversationId.capture())(any[HeaderCarrier]())
-    }
+    "must treat file having size > maxNormalFileSizeBytes as large file" in {
+      forAll(arbitrary[SubmissionDetails]) { submissionDetails =>
```
Comment (oarshad) [inline test/controllers/submission/SubmissionControllerSpec.scala:69]: same as above - and also every where if we passing implicit arbitraries within `forAll`

File: test/services/submission/SDESFileMetadataServiceSpec.scala [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+  "SDESMetaDataServiceSpec" - {
+    "compileMetaData" - {
+
+      "must compile metadata for an organisation" in {
+        forAll(
+          arbitrary[ResponseDetail],
+          arbitrary[ConversationId],
+          arbitrary[String].suchThat(_.nonEmpty)
+        ) { (responseDetail, conversationId, fileName) =>
```
Comment (oarshad) [inline test/services/submission/SDESFileMetadataServiceSpec.scala:43]:
```
      implicit val nonEmptyString: Arbitrary[String] = Arbitrary(Gen.alphaNumStr.suchThat(_.nonEmpty))

      "must compile metadata for an organisation" in {
        forAll{ (responseDetail: ResponseDetail, conversationId: ConversationId, fileName: String) =>
        ```

File: test/services/submission/SDESServiceSpec.scala [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+      bind[Clock].toInstance(fixedClock)
+    )
+    .build()
+
+  private val sdesService = app.injector.instanceOf[SDESService]
+
+  "SDESService" - {
+    "sendFileNotification" - {
+      "must send file submission notification to SDES connector and return success status when connector returns a Right" in {
+        forAll(arbitrary[ResponseDetail], arbitrary[SubmissionDetails]) { (responseDetail, submissionDetails) =>
```
Comment (oarshad) [inline test/services/submission/SDESServiceSpec.scala:60]: same as above

**Review by oarshad — CHANGES_REQUESTED** [review]
Looks good overall - some minor suggestions

File: app/models/xml/XmlHandler.scala [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+import java.net.URL
+import scala.xml.Elem
+
+class XmlHandler {
+
+  def load(url: String): Elem =
+    new scala.xml.factory.XMLLoader[scala.xml.Elem] {
+      override def adapter = new scala.xml.parsing.NoBindingFactoryAdapter
+    }.load(new URL(url))
```
[the remark this answers is not in the document]
Comment (koladeadewuyi) [reply app/models/xml/XmlHandler.scala]: We're using java 11 which does not include `URI`.
It is now wrapped in a `Try` to handle any exception which may be thrown.

**Review by oarshad — APPROVED** [review]

**Pull request merged**